envoy/config/grpc_mux.h

```diff
@@ -105,6 +110,32 @@ class GrpcMux {
 
   virtual void requestOnDemandUpdate(const std::string& type_url,
                                      const absl::flat_hash_set<std::string>& for_update) PURE;
+
+  // Unified mux interface starts here
+  // Updates the list of resource names watched by the given watch. If an added name is new across
+  // the whole subscription, or if a removed name has no other watch interested in it, then the
+  // subscription will enqueue and attempt to send an appropriate discovery request.
+  virtual void updateWatch(const std::string& type_url, Watch* watch,
```

Comment:
`Watch` wasn't previously part of the pure gRPC mux interface? How come this is now exposed? Seems somewhat of an implementation detail.

Reply:
It sort of was -- `GrpcMux::addWatch` returns an `Config::GrpcMuxWatchPtr`; `GrpcMuxWatch` is an interface for a wrapper around `Watch`. Instead of relying on a wrapper, the idea is for the subscription to call `updateWatch`directly, bypassing the wrapper. There's a forward declaration of `Watch` here: https://github.com/envoyproxy/envoy/pull/17352/files#diff-e9dc2eefa8ac985ee3728f41198533907528fcbfebf25d296450ee81f9c0eb59R37.

Reply:
Yeah, but do we gain something from this? Ideally we have an abstract minimal interface in `include/envoy` that covers the desired behavior, which I think is `GprcMuxWatch`. Then we have concrete instantiations, which is `Watch`. We can mock the abstract interface. This is a bit cleaner than forward declaring something form the implementation.

Reply:
It's a weird interface though, it only has a single method `update()`, that acts as a facade for mux's `updateWatch()`. I feel it's quite bit of scaffolding (and we additionally wrap the instance of GrpcMuxWatch into a unique_ptr) for something fairly trivial, which makes it harder to understand/maintain. 
If you fell strongly about it, I'll redo this with `GrpcMuxWatch` interface instead.

Reply:
I think I prefer an abstract interface which owns its own method. Logically a watch is created and can be implemented. We shouldn't couple with the factory at the interface level. In addition, the `removeWatch` can then be skipped by use of RAII.

Reply:
done.